core/snippets/snippets.py

```diff
@@ -52,128 +63,155 @@ def get_setting_dir():
-        insertion_map = {}
-        insertions_phrase_map = {}
-        wrapper_map = {}
-
-        # Assign global snippets to all languages
-        for lang_super in ["_", lang]:
-            snippets, insertions, insertions_phrase, wrappers = create_lists(
-                lang,
-                lang_super,
-                language_to_snippets.get(lang_super, []),
-            )
-            snippets_map.update(snippets)
-            insertion_map.update(insertions)
-            insertions_phrase_map.update(insertions_phrase)
-            wrapper_map.update(wrappers)
-
-        ctx.lists.update(
-            {
-                "user.snippet": insertion_map,
-                "user.snippet_with_phrase": insertions_phrase_map,
-                "user.snippet_wrapper": wrapper_map,
-            }
-        )
+def get_preferred_snippet(snippets: list[Snippet]) -> Snippet:
+    lang: Union[str, set[str]] = actions.code.language()
+    languages = [lang] if isinstance(lang, str) else lang
 
+    # First try to find a snippet matching the active language
+    for snippet in snippets:
+        if snippet.languages:
+            for snippet_lang in snippet.languages:
+                if snippet_lang in languages:
+                    return snippet
 
-def get_snippets() -> list[Snippet]:
-    files = glob.glob(f"{SNIPPETS_DIR}/**/*.snippet", recursive=True)
+    # Then look for a global snippet
+    for snippet in snippets:
+        if not snippet.languages:
+            return snippet
 
-    if get_setting_dir():
-        files.extend(glob.glob(f"{get_setting_dir()}/**/*.snippet", recursive=True))
+    raise ValueError(f"Snippet not available for language '{lang}'")
 
-    result = []
 
-    for file in files:
-        result.extend(create_snippets_from_file(file))
+def split_wrapper_snippet_name(name: str) -> tuple[str, str]:
+    index = name.rindex(".")
+    return name[:index], name[index + 1]
 
-    return result
+
+def to_wrapper_snippet(snippet: Snippet, variable_name) -> WrapperSnippet:
+    """Get wrapper snippet named <name>"""
+    var = snippet.get_variable_strict(variable_name)
+    return WrapperSnippet(
+        snippet.body,
+        var.name,
+        var.wrapper_scope,
+        snippet.languages,
+    )
 
 
-def group_by_language(snippets: list[Snippet]) -> dict[str, list[Snippet]]:
-    result = defaultdict(list)
+def update_snippets():
+    global snippets_map
+
+    snippets = get_snippets_from_files()
+    name_to_snippets: dict[str, list[Snippet]] = {}
+    language_to_lists: dict[str, SnippetLists] = {}
+
     for snippet in snippets:
-        if snippet.languages is not None:
-            for lang in snippet.languages:
-                result[lang].append(snippet)
-        else:
-            result["_"].append(snippet)
-    return result
+        # Map snippet names to actual snippets
+        name_to_snippets.setdefault(snippet.name, [])
+        name_to_snippets[snippet.name].append(snippet)
```

Comment:
Sorry, forgot to mention that the return value of `setdefault` is also useful!

```suggestion
        name_to_snippets.setdefault(snippet.name, []).append(snippet)
```